Fix INSERT query performance on compressed chunk

sb230132 opened the pull request:
Fix INSERT query performs on compressed chunk

The INSERT query with ON CONFLICT on compressed chunk does a
heapscan to verify unique constraint violation. This patch
improves the performance by doing an indexscan on compressed
chunk, to fetch matching records based on segmentby columns.
These matching records are inserted into uncompressed chunk,
then unique constraint violation is verified.

Since index on compressed chunk has only segmentby columns,
we cannot do a point lookup considering orderby columns as well.
This patch thus will result in decompressing matching
records only based on segmentby columns.

Fixes #6063
Fixes #5801

sb230132 commented:
Changes in compression_conflicts_iso are expected. When i ran the same test on postgres without hypertables, i see same results.
[pg_ddl_iso.sql.zip](https://github.com/timescale/timescaledb/files/12584312/pg_ddl_iso.sql.zip)
[pg_ddl_iso.out.zip](https://github.com/timescale/timescaledb/files/12584314/pg_ddl_iso.out.zip)

jnidzwetzki commented on tsl/src/compression/compression.c:
```diff
@@ -2005,82 +2130,120 @@ decompress_batches_for_insert(ChunkInsertState *cis, Chunk *chunk, TupleTableSlo
+				if (!heap_attisnull(compressed_tuple, attno, decompressor.in_desc))
+				{
+					valid = false;
+					break;
```
Could you add a test for this line?

sb230132 replied:
tsl/test/sql/compression_conflicts.sql has test which will cover these lines.

jnidzwetzki replied:
It seems Codecov still does not see any coverage here. Which SQL statement from the test should cover these lines?

jnidzwetzki commented:
@sb230132 If I understood the changes in the test output correctly, this PR is not only about performance - it also fixes wrong query results. I suggest adding this fact to the commit message.

jnidzwetzki commented on tsl/src/compression/compression.c:
```diff
@@ -1836,6 +1836,62 @@ tsl_compressed_data_out(PG_FUNCTION_ARGS)
+	 * locking the compressed chunk here
+	 */
+	table_close(compressed_chunk_rel, NoLock);
+	table_close(uncompressed_chunk_rel, NoLock);
```
Is it intended to keep the lock of the `uncompressed_chunk_rel` too? If so, could you add a comment why this is required?

sb230132 replied:
There is a comment just above.

jnidzwetzki replied:
This comment only covers the `compressed_chunk_rel`. If the same applies to the `uncompressed_chunk_rel` it should be added to the comment.

jnidzwetzki commented on tsl/src/compression/compression.h:
```diff
@@ -391,6 +391,13 @@ consumeCompressedData(StringInfo si, int bytes)
+/* build scankeys for segmentby columns */
+#define SEGMENTBY_KEYS (1 << 1)
+/* build scankeys for orderby columns */
+#define ORDERBY_KEYS (1 << 2)
```
Do you need this define? It seems it is never used in the code.

sb230132 replied:
Removed ORDERBY_KEYS as it is not used anywhere

jnidzwetzki commented on tsl/src/compression/compression.c:
```diff
@@ -2005,82 +2130,120 @@ decompress_batches_for_insert(ChunkInsertState *cis, Chunk *chunk, TupleTableSlo
+			write_logical_replication_msg_decompression_end();
+
+			TM_FailureData tmfd;
+			TM_Result result pg_attribute_unused();
```
This code was already present in the old code. Could you please add a comment to explain why it is safe to ignore the result of the operation?

jnidzwetzki replied:
It seems you added also an `assert`. My question was more _why_ is it safe to ignore the result of the operation? Maybe you could add this information to the comment.

akuzm commented on tsl/src/compression/api.c:
```diff
@@ -1017,6 +1017,7 @@ tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
 {
 	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
 	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+	Chunk *compressed_chunk = ts_chunk_get_by_id(uncompressed_chunk->fd.compressed_chunk_id, true);
```
This lookup is expensive, is it enough to get the relid here? There's a less heavy function `ts_chunk_get_relid` for that, or maybe `ts_chunk_get_formdata`.

Same comment for another usage below.

sb230132 replied:
Fixed as you suggested. Help me understand why ```ts_chunk_get_by_id``` is a costly operation ?

```ts_chunk_get_by_id``` does following:
1. Build scankey on chunk id.
2. Take AccessShareLock on CHUNK catalog table.
3. Scan the index and fetch the required row.

What you suggested is:
1. Call ```ts_chunk_get_relid``` to get chunk relid.
2. Call ```ts_chunk_get_by_relid``` to get the chunk from relid.

```ts_chunk_get_by_relid``` does following:
1. Build scankey on schem, chunk name.
2. Take AccessShareLock on CHUNK catalog table.
3. Scan the index and fetch the required row.

akuzm commented on tsl/src/compression/compression.c:
```diff
@@ -1986,9 +2042,78 @@ create_segment_filter_scankey(RowDecompressor *decompressor, char *segment_filte
+		operator= get_opfamily_member(opfamily, optype, optype, BTEqualStrategyNumber);
+		regop = get_opcode(operator);
+
+		if (attoff < indnkeyatts - 1)
+		{
+			/* Initialize segmentby scankeys. */
+			if (key_flags & SEGMENTBY_KEYS)
+			{
+				/* get index attribute name */
+				Form_pg_attribute attr = TupleDescAttr(idxrel->rd_att, attoff);
```
Can we end up with index scan with no keys here? Probably better to keep the seq scan for this case, because a full index scan w/o keys is just needlessly slower.

sb230132 replied:
I don't understand what you mean here ?
By default there is always an index with segmentby columns on compressed chunk.
Index Oid is fetched from get_compressed_chunk_index() which will always check for presence of segmentby columns in the index, else this function will return InvalidOid.

akuzm replied:
I mean, can `key_flags & SEGMENTBY_KEYS` be false? That would mean we're using an index scan, but without any index conditions.

sb230132 replied:
Fixed.

sb230132 committed:
Fix INSERT query performance on compressed chunk

The INSERT query with ON CONFLICT on compressed chunk does a
heapscan to verify unique constraint violation. This patch
improves the performance by doing an indexscan on compressed
chunk, to fetch matching records based on segmentby columns.
These matching records are inserted into uncompressed chunk,
then unique constraint violation is verified.

Since index on compressed chunk has only segmentby columns,
we cannot do a point lookup considering orderby columns as well.
This patch thus will result in decompressing matching
records only based on segmentby columns.

Fixes #6063
Fixes #5801

svenklemm reviewed (changes requested):
This seems to not handle attribute numbers correctly and use them directly from the hypertable instead of the chunks. In addition this seems to completely change the operator handling and always use opfamily operators without checking for compatibility.

horzsolt commented:
Closing this one, @antekresic is working on some other PRs which will replace this one.

mgagliardo91 commented:
@horzsolt @antekresic  would you mind posting a link to the new PRs when they are created?

mblsf commented:
@horzsolt @antekresicv I would also like to know these PRs

svenklemm commented:
@mgagliardo91 @mblsf some fixes already went into 2.12.0 with https://github.com/timescale/timescaledb/pull/6081 but there are more followup PRs coming
